RecoEgamma/EgammaTools/interface/HGCalShowerShapeHelper.h

```diff
@@ -0,0 +1,131 @@
+#include "DataFormats/FWLite/interface/ESHandle.h"
+#include "DataFormats/ForwardDetId/interface/HGCEEDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+#include "DataFormats/HGCRecHit/interface/HGCRecHit.h"
+#include "DataFormats/Math/interface/LorentzVector.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecHit.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecHitFraction.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "Geometry/CaloTopology/interface/HGCalTopology.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h"
+#include "RecoParticleFlow/PFClusterProducer/interface/InitialClusteringStepBase.h"
+
+#include <CLHEP/Vector/LorentzVector.h>
+
+#include <algorithm>
+#include <cstdlib>
+#include <iostream>
+#include <map>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <TMatrixD.h>
+#include <TVectorD.h>
+
+#include <Math/Point3D.h>
+#include <Math/Point3Dfwd.h>
+
+class HGCalShowerShapeHelper {
+private:
+  // Good to filter/compute/store this stuff beforehand as they are common to the shower shape variables.
+  // No point in filtering, computing layer-wise centroids, etc. for each variable again and again.
+  // Once intitialized, one can the calculate different variables one after another for a given object.
+  // If a different set of preselections (E, ET, etc.) is required for a given object, then reinitialize using initPerObject(...).
+
+  // In principle should consider the HGCalHSi and HGCalHSc hits (leakage) also.
+  // Can have subdetector dependent thresholds and layer selection.
+  // To be implemented.
+
+  double minHitE_;
+  double minHitET_;
+  double minHitET2_;
+  int minLayer_;
+  int maxLayer_;
+  int nLayer_;
+  DetId::Detector subDet_;
+
+  hgcal::RecHitTools recHitTools_;
+
+  std::unordered_map<uint32_t, const reco::PFRecHit *> pfRecHitPtrMap_;
+  std::vector<std::pair<DetId, float> > hitsAndFracs_;
+  std::vector<double> hitEnergies_;
+  std::vector<double> hitEnergiesWithFracs_;
+
+  ROOT::Math::XYZVector centroid_;
+  std::vector<double> layerEnergies_;
+  std::vector<ROOT::Math::XYZVector> layerCentroids_;
+
+  void setPFRecHitPtrMap(const std::vector<reco::PFRecHit> &recHits);
+
+  void setFilteredHitsAndFractions(const std::vector<std::pair<DetId, float> > &hitsAndFracs);
+
+public:
+  static constexpr double kLDWaferCellSize_ = 0.698;
+  static constexpr double kHDWaferCellSize_ = 0.465;
```

Comment:
can these be taken from somewhere else in HGCAL ?

Reply:
I'd have to check it.

Reply:
I believe something _similar_ can be derived from the geometry. We do have the ability in Fireworks to display the single cells. There's a method `getNewCorners` from the `HGCalGeoemtry` [see here](https://cmssdt.cern.ch/dxr/CMSSW/source/Geometry/HGCalGeometry/src/HGCalGeometry.cc#356) that could be used. But, in this case, I'd ask Sunanda for the exact meaning of the returned numbers.
Also, that would imply removing the `constexpr`, I believe.

Reply:
`constexpr` is not particularly useful here, since these are used after an `if` condition based on runtime checks.

Anyway, if these numbers are only used here, it makes sense to keep them in this class.